x-pack/plugins/lens/public/indexpattern_plugin/indexpattern.tsx

```diff
@@ -95,76 +124,49 @@ export function IndexPatternDataPanel(props: DatasourceDataPanelProps<IndexPatte
   );
 }
 
-export type IndexPatternDimensionPanelProps = DatasourceDimensionPanelProps & {
-  state: IndexPatternPrivateState;
-  setState: (newState: IndexPatternPrivateState) => void;
-};
+export function columnToOperation(column: IndexPatternColumn) {
+  const { dataType, label, isBucketed, operationId } = column;
+  return {
+    id: operationId,
+    label,
+    dataType,
+    isBucketed,
+  };
+}
 
-export function IndexPatternDimensionPanel(props: IndexPatternDimensionPanelProps) {
-  const fields = props.state.indexPatterns[props.state.currentIndexPatternId].fields;
-  const columns: IndexPatternColumn[] = fields.map((field, index) => ({
-    operationId: `${index}`,
-    label: `Value of ${field.name}`,
-    dataType: field.type as DataType,
-    isBucketed: false,
+type UnwrapPromise<T> = T extends Promise<infer P> ? P : T;
+type InferFromArray<T> = T extends Array<infer P> ? P : T;
 
-    operationType: 'value' as OperationType,
-    sourceField: field.name,
-  }));
+function addRollupInfoToFields(
+  indexPattern: InferFromArray<Exclude<UnwrapPromise<ReturnType<typeof getIndexPatterns>>, void>>
+): IndexPattern {
+  if (!indexPattern.typeMeta || typeof indexPattern.fields === 'string') {
```

Comment:
I didn't expect indexPattern.fields to be ever a string so I looked into the origin of this type. It turns out in `x-pack/plugins/lens/public/indexpattern_plugin/loader.ts` Object.assign is used to build this object which for some reason mixes the interfaces of its parameters. An object spread doesn't do this:
```ts
        return {
          ...attributes,
          timeFieldName: attributes.timeFieldName || undefined,
          id,
          type,
          title: attributes.title,
          fields: (JSON.parse(attributes.fields) as IndexPatternField[]).filter(
            ({ type: fieldType, esTypes }) =>
              fieldType !== 'string' || (esTypes && esTypes.includes('keyword'))
          ),
          typeMeta: attributes.typeMeta
            ? (JSON.parse(attributes.typeMeta) as SavedRollupInfo)
            : undefined,
        }
```

Then `indexPattern.fields` is `IndexPatternField[]` all the time and the code here can be simplified to

```ts
  if (!indexPattern.typeMeta) {
    return indexPattern;
  }

  const typeMeta = indexPattern.typeMeta;

  const aggs = Object.keys(typeMeta.aggs);

  const newFields = [...indexPattern.fields];
  newFields.forEach((field, index) => {
    const restrictionsObj: IndexPatternField['rollupRestrictions'] = {};
    aggs.forEach(agg => {
      if (typeMeta.aggs[agg] && typeMeta.aggs[agg][field.name]) {
        restrictionsObj[agg] = typeMeta.aggs[agg][field.name];
      }
    });
// ....
```

Reply:
This is what I was trying to get to, I'll update this.